File: sdks/php/tests/ChannelsResourceTest.php
<?php

declare(strict_types=1);

namespace Linktor\Tests;

use Linktor\LinktorClient;
use Linktor\Resources\ChannelsResource;
use Linktor\Types\Channel;
use Linktor\Types\ConnectResult;
use PHPUnit\Framework\TestCase;

/**
 * The HttpClient already unwraps the `{ success, data }` envelope, so the
 * resource sees the inner `data`. These tests mock LinktorClient (whose public
 * get/post/put methods return that unwrapped `data`) to prove the resource
 * shapes it into the right wire types.
 */
class ChannelsResourceTest extends TestCase
{
    private function channelData(): array
    {
        return [
            'id' => 'ch_1',
            'tenant_id' => 't_1',
            'type' => 'whatsapp_unofficial',
            'name' => 'WhatsApp',
            'enabled' => true,
            'connection_status' => 'connecting',
            'config' => ['phone_number_id' => '123', 'access_token' => '__redacted__'],
            'created_at' => '2026-07-24T10:00:00Z',
            'updated_at' => '2026-07-24T10:00:00Z',
        ];
    }

    public function testConnectSurfacesQrCode(): void
    {
        // Inner `data` for POST /channels/{id}/connect, post-envelope-unwrap.
        $unwrapped = [
            'channel' => $this->channelData(),
            'qr_code' => 'Q',
            'expires_in' => 60,
        ];

        $client = $this->createMock(LinktorClient::class);
        $client->expects($this->once())
            ->method('post')
            ->with('/channels/ch_1/connect', [])
            ->willReturn($unwrapped);

        $channels = new ChannelsResource($client);
        $result = $channels->connect('ch_1');

        $this->assertInstanceOf(ConnectResult::class, $result);
        $this->assertSame('Q', $result->qrCode);
        $this->assertSame(60, $result->expiresIn);
        $this->assertInstanceOf(Channel::class, $result->channel);
        $this->assertSame('ch_1', $result->channel->id);
        $this->assertSame('connecting', $result->channel->connectionStatus);
    }

    public function testRequestPairCodeSendsPhoneAndSurfacesPairCode(): void
    {
        $unwrapped = [
            'channel' => $this->channelData(),
            'pair_code' => 'ABCD-1234',
            'expires_in' => 120,
        ];

        $client = $this->createMock(LinktorClient::class);
        $client->expects($this->once())
            ->method('post')
            ->with('/channels/ch_1/pair', ['phone_number' => '+15551234567'])
            ->willReturn($unwrapped);

        $channels = new ChannelsResource($client);
        $result = $channels->requestPairCode('ch_1', '+15551234567');

        $this->assertSame('ABCD-1234', $result->pairCode);
        $this->assertSame(120, $result->expiresIn);
    }

    public function testCreateSendsCredentials(): void
    {
        $input = [
            'type' => 'whatsapp',
            'name' => 'WhatsApp',
            'config' => ['phone_number_id' => '123'],
            'credentials' => ['access_token' => 'secret'],
        ];

        $client = $this->createMock(LinktorClient::class);
        $client->expects($this->once())
            ->method('post')
            ->with('/channels', $input)
            ->willReturn($this->channelData());

        $channels = new ChannelsResource($client);
        $channel = $channels->create($input);

        $this->assertInstanceOf(Channel::class, $channel);
        $this->assertSame('ch_1', $channel->id);
        $this->assertTrue($channel->enabled);
        // Credentials are write-only: they must never surface on the response.
        $this->assertArrayNotHasKey('access_token', $channel->config);
        $this->assertSame('__redacted__', $channel->config['access_token'] ?? null);
    }

    public function testUpdateUsesPut(): void
    {
        $client = $this->createMock(LinktorClient::class);
        $client->expects($this->once())
            ->method('put')
            ->with('/channels/ch_1', ['name' => 'Renamed'])
            ->willReturn(['id' => 'ch_1', 'tenant_id' => 't_1', 'type' => 'whatsapp', 'name' => 'Renamed']);

        $channels = new ChannelsResource($client);
        $channel = $channels->update('ch_1', ['name' => 'Renamed']);

        $this->assertSame('Renamed', $channel->name);
    }

    public function testListReturnsChannelArray(): void
    {
        $client = $this->createMock(LinktorClient::class);
        $client->expects($this->once())
            ->method('get')
            ->with('/channels?type=whatsapp')
            ->willReturn([$this->channelData(), $this->channelData()]);

        $channels = new ChannelsResource($client);
        $list = $channels->list(['type' => 'whatsapp']);

        $this->assertCount(2, $list);
        $this->assertContainsOnlyInstancesOf(Channel::class, $list);
    }
}

File: sdks/java/src/test/java/io/linktor/resources/ChannelsResourceTest.java
package io.linktor.resources;

import com.sun.net.httpserver.HttpServer;
import io.linktor.types.Channel;
import io.linktor.utils.HttpClient;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicReference;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Focused tests proving the ChannelsResource speaks the real wire contract:
 * the {@code {success,data}} envelope is unwrapped, connect() surfaces qr_code,
 * and create() sends write-only credentials.
 */
class ChannelsResourceTest {

    private HttpServer server;
    private ChannelsResource channels;
    private final AtomicReference<String> lastRequestBody = new AtomicReference<>();
    private final Map<String, String> cannedResponses = new HashMap<>();

    @BeforeEach
    void setUp() throws IOException {
        server = HttpServer.create(new InetSocketAddress("127.0.0.1", 0), 0);
        server.createContext("/", exchange -> {
            byte[] reqBytes = exchange.getRequestBody().readAllBytes();
            lastRequestBody.set(new String(reqBytes, StandardCharsets.UTF_8));

            String path = exchange.getRequestURI().getPath();
            String method = exchange.getRequestMethod();
            String key = method + " " + path;
            String body = cannedResponses.getOrDefault(key, "{\"success\":true,\"data\":null}");

            byte[] out = body.getBytes(StandardCharsets.UTF_8);
            exchange.getResponseHeaders().add("Content-Type", "application/json");
            exchange.sendResponseHeaders(200, out.length);
            exchange.getResponseBody().write(out);
            exchange.close();
        });
        server.start();

        String baseUrl = "http://127.0.0.1:" + server.getAddress().getPort();
        HttpClient http = new HttpClient(baseUrl, "lk_test", null, 5, 1);
        channels = new ChannelsResource(http);
    }

    @AfterEach
    void tearDown() {
        server.stop(0);
    }

    @Test
    void connectSurfacesQrCodeFromEnvelope() throws Exception {
        cannedResponses.put("POST /channels/c1/connect",
                "{\"success\":true,\"data\":{\"channel\":{\"id\":\"c1\",\"type\":\"whatsapp\","
                        + "\"connection_status\":\"connecting\",\"enabled\":true},"
                        + "\"qr_code\":\"Q\",\"expires_in\":60}}");

        Channel.ConnectResult result = channels.connect("c1");

        assertNotNull(result);
        assertEquals("Q", result.getQrCode());
        assertEquals(60, result.getExpiresIn());
        assertNotNull(result.getChannel());
        assertEquals("c1", result.getChannel().getId());
        assertEquals(Channel.ConnectionStatus.CONNECTING, result.getChannel().getConnectionStatus());
        assertEquals(Channel.ChannelType.WHATSAPP, result.getChannel().getType());
    }

    @Test
    void requestPairCodeSendsPhoneAndSurfacesPairCode() throws Exception {
        cannedResponses.put("POST /channels/c1/pair",
                "{\"success\":true,\"data\":{\"channel\":{\"id\":\"c1\"},\"pair_code\":\"ABCD-1234\"}}");

        Channel.ConnectResult result = channels.requestPairCode("c1", "+5511999999999");

        assertEquals("ABCD-1234", result.getPairCode());
        assertTrue(lastRequestBody.get().contains("\"phone_number\""));
        assertTrue(lastRequestBody.get().contains("+5511999999999"));
    }

    @Test
    void createSendsCredentialsAndUnwrapsChannel() throws Exception {
        cannedResponses.put("POST /channels",
                "{\"success\":true,\"data\":{\"id\":\"c9\",\"type\":\"whatsapp\",\"name\":\"WA\","
                        + "\"enabled\":true,\"connection_status\":\"disconnected\","
                        + "\"config\":{\"access_token\":\"__redacted__\"}}}");

        Map<String, String> creds = new HashMap<>();
        creds.put("access_token", "super-secret-token");
        Channel.CreateChannelInput input = Channel.CreateChannelInput.builder()
                .type(Channel.ChannelType.WHATSAPP)
                .name("WA")
                .credentials(creds)
                .build();

        Channel.ChannelModel created = channels.create(input);

        // Response is unwrapped from the envelope.
        assertEquals("c9", created.getId());
        assertEquals(Channel.ConnectionStatus.DISCONNECTED, created.getConnectionStatus());

        // Credentials were sent on the wire (snake_case, write-only).
        String sent = lastRequestBody.get();
        assertTrue(sent.contains("\"credentials\""), "credentials missing: " + sent);
        assertTrue(sent.contains("\"access_token\""), "access_token missing: " + sent);
        assertTrue(sent.contains("super-secret-token"), "secret value missing: " + sent);
        assertTrue(sent.contains("\"whatsapp\""), "type not serialized as wire enum: " + sent);
    }

    @Test
    void listUnwrapsPlainArray() throws Exception {
        cannedResponses.put("GET /channels",
                "{\"success\":true,\"data\":[{\"id\":\"a\"},{\"id\":\"b\"}],\"meta\":{}}");

        List<Channel.ChannelModel> result = channels.list();

        assertEquals(2, result.size());
        assertEquals("a", result.get(0).getId());
        assertEquals("b", result.get(1).getId());
    }
}
